src/util.ts

```diff
@@ -343,3 +345,82 @@ export const escapeCommand = (args: string[]) => {
     const escapedCommand = args.map((arg) => escapeArg(arg)).join(' ');
     return IS_QUOTED_REGEXP.test(escapedCommand) ? escapedCommand : `'${escapedCommand}'`;
 };
+
+/**
+ * Returns a list of all detected Android and iOS devices currently connected to the host. This includes Android
+ * emulators running on the host.
+ *
+ * @param options If the `frida` option is set to `true`, this function will use frida to detect the devices rather than
+ *   try to detect them using platform-specific tools (such as `adb` and `pymobiledevice3`).
+ */
+export const listDevices = async (options?: {
+    frida?: boolean;
+}): Promise<
+    {
+        /** A unique id that identifies the device to the platform tools and frida. */
+        id: string;
+        /**
+         * The name or model of the device (adb sometimes can not find that information easily, so it might be
+         * undefined)
+         */
+        name?: string;
+        /** The platform the device is running on. */
+        platform: 'android' | 'ios';
+    }[]
+> => {
+    if (options?.frida)
+        return await frida
+            .enumerateDevices()
+            .then((devices) => devices.filter((device) => device.type === 'usb'))
+            .then((devices) =>
+                Promise.all(
+                    devices.map(async (device) => {
+                        const params = await device.querySystemParameters();
+                        return {
+                            id: device.id,
+                            name: device.name,
+                            platform: params.platform === 'darwin' ? 'ios' : 'android',
+                        };
+                    })
+                )
+            );
+
+    const venv = getVenv(venvOptions);
+    const python = async (...args: Parameters<Awaited<typeof venv>>) => (await venv)(...args);
+
+    const androidDevices = await runAndroidDevTool('adb', ['devices', '-l']).then(({ stdout }) => {
+        const lines = stdout.split('\n');
+        if (lines[0]?.includes('List of devices attached'))
+            return lines
+                .slice(1, -1)
+                .map((line) => {
+                    const device = line.split(/\s+/);
+
+                    // Only return devices which are actually connected (possible vaules are 'device', 'offline' and 'no', see https://developer.android.com/tools/adb#devicestatus)
+                    if (device[1] !== 'device') return undefined;
+
+                    return device[0] && device[1]
+                        ? {
+                              id: device[0],
+                              name: line.match(/model:(.+)? /)?.[1],
+                              platform: 'android' as const,
+                          }
+                        : undefined;
+                })
+                .filter(Boolean) as { id: string; name?: string; platform: 'android' }[];
+        throw new Error('Could not list android devices: unexpected output from adb devices');
```

Comment:
Android throws on unexpected output, iOS ignores it silently. I think ignoring silently is the way to go. If you're only interested in analysing one platform, the setup for the other may be broken and that shouldn't be a problem.